libs/block_generator/block_generator.cpp

```diff
@@ -0,0 +1,289 @@
+#include "block_generator.hpp"
+
+#include <boost/algorithm/hex.hpp>
+#include <boost/json.hpp>
+#include <boost/json/array.hpp>
+#include <boost/json/serialize.hpp>
+#include <boost/json/stream_parser.hpp>
+#include <cstddef>
+#include <cstdint>
+#include <expected>
+#include <iterator>
+#include <optional>
+#include <sstream>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "json_validator.hpp"
+#include "schema.def"
+#include "zkevm_framework/data_types/account_block.hpp"
+#include "zkevm_framework/data_types/base.hpp"
+#include "zkevm_framework/data_types/block.hpp"
+#include "zkevm_framework/data_types/block_header.hpp"
+#include "zkevm_framework/data_types/message.hpp"
+#include "zkevm_framework/data_types/transaction.hpp"
+
+namespace {
+    std::unordered_map<size_t, data_types::Transaction> transaction_map;
+
+    // Iterate over keys in the entity and call handler for each of them
+    template<typename Callback, typename Result>
+    void iterate_object(const boost::json::object &entity, Result &res, Callback key_handler) {
+        for (const auto &elem : entity) {
+            key_handler(elem.key_c_str(), elem.value(), res);
+        }
+    }
+
+    size_t get_number(const boost::json::value &json_value) {
+        if (json_value.is_int64()) {
+            return json_value.as_int64();
+        }
+        return json_value.as_uint64();
+    }
+
+    // Transform hex string into Address or bytes
+    template<typename OutputContainer>
+    void get_bytes(const boost::json::value hex_value, OutputContainer &result) {
+        std::string hex_string = hex_value.as_string().c_str();
+        std::vector<uint8_t> bytes;
+        // Skip 0x prefix if exists
+        if (hex_string[0] == '0' && hex_string[1] == 'x') {
+            hex_string.erase(0, 2);
+        }
+        boost::algorithm::unhex(hex_string, std::back_inserter(bytes));
+        if constexpr (std::is_same_v<OutputContainer, data_types::Address>) {
+            std::copy(bytes.begin(), bytes.end(), result.begin());
+        } else {
+            // Store the data of a transaction, could be any size
+            result.clear();
+            for (uint8_t b : bytes) {
+                result.push_back(std::byte(b));
+            }
+        }
+    }
+
+    void handle_transaction(const std::string &key, const boost::json::value &json_value,
+                            data_types::Transaction &transaction) {
+        if (key == "id") {
+            transaction.m_id = get_number(json_value);
```

Comment:
One thing which bothers me a bit here is that you use Transaction fields directly (and other data types also). Probably these fields will be private when we will come up with consistent interfaces for Data Types. So we need to be ready to change this code as well.

But it's okay for now

Reply:
Well, that was the only interface for now to set fields. I suppose we will anyway need some setters for fields, because the constructor is not convenient due to the number of arguments.